### ISSUE TITLE
All Examples should use the same vmlinux.h file

### ISSUE THREAD

**Issue by mimullin-bbry** [libbpf/libbpf-rs#143]
No need for every example to have it's own vmlinux.h file.  

Put one vmlinux_505.h inside of a ./examples/header/ directory and have all examples' vmlinux.h files soft link to that. (or something similar)

**Comment by DevasiaThomas** [comment]
@mimullin-bbry I had a few questions about this :). does the vmlinux.h file vary across OS distros for the same kernel version? Is it better to compile with a vmlinux.h from a higher kernel version or lower one when considering CO-RE applications?

**Comment by DevasiaThomas** [comment]
Hmm, I commented on that PR before I read this, because I was worried about CO-RE.  I'll continue the conversation there. #144

**Comment by anakryiko** [comment]
> > Can the vmlinux.h file vary across OS distros for the same kernel version?
> 
> Yes. This seen when using RHEL or RHEL derivative kernels. RHEL backports a lot of bpf code (which is why you see some #ifdef RHEL stuff in tracee) to it's old kernels. Along with these backports, you'll see at minimum the additional bpf helpers.
> 
> > Is it better to compile with a vmlinux.h from a higher kernel version or lower one when considering CO-RE applications?

Generally newer kernels have more features so it makes sense to use newest vmlinux.h. You'll still have to take into considerations all the incompatible changes in old kernels. 

> 
> I don't have the full technical picture to give you a definitive answer, but if you see the the PR associated with this Issue, you'll notice a major break for the task_struct::state field. With the state field changing in 5.14, BPF code that uses the state field will not compile on kernel 5.14+, and bpf code that uses the new __state field will not pass the verifier on kernels < 5.14. (see https://github.com/libbpf/libbpf-rs/pull/144/files).
> 
> The end result of both of these answers put together is that even with CORE, you need to focus your project on a target distribution-version-kernel triple (ie ubuntu-20.04-kernel5.8aws) that has the bare minimum features you need, and as you progress to other distributions+versions+kernels you will need to make hacks (sometimes requiring recompilation, see the state vs __state).

You most certainly don't want to base this on distro-version-kernel triple if at all avoidable. BPF CO-RE provides various mechanism to both detect which specific set of features/fields kernel provides, as well as an ability to handle that within a single compiled BPF CO-RE application. See https://nakryiko.com/posts/bpf-core-reference-guide/#handling-incompatible-field-and-type-changes.

> 
> I realize this is a very unsatisfying answer, which might lead you to be thinking "why don't I just use BCC instead of libbpf + CORE". But I believe that libbpf is making progress and will be the ultimately better technology.

It's just simply not true. BPF CO-RE does provide ways to still handle all that. Not all breaking changes can be handled automatically, so you'll need to track such kernel changes and adopt your code, but you don't need to compile multiple version of your BPF application and certainly don't have to make it distro+kernel specific. See my comment on https://github.com/libbpf/libbpf-rs/pull/144#issuecomment-965834731.

**Comment by mimullin-bbry** [comment]
@anakryiko Thank you for helping to educate us.  BPF has been lacking some good documentation, and your recent posts have helped spread some good knowledge.  EG I had no knowledge of bpf_core_field_exists() until yesterday.

A lot of my suggestion to use older kernels rather than newer kernels comes to missing bpf-helper functions and missing features such as bpf_ringbuffers. And of course, the dread glibc issues (though containerization takes this pain away). It's painful to develop a bpf application on Kernel 5.10 using features such as ringbuffers and bpf_d_path, linking against glibc 2.33, to then realize the code needs to run on 5.4 glibc 2.28.

IMHO it's easier to get the program going on 5.4 and then pull features into the app for targetting newer kernels/distros.

> It's just simply not true.

I apologize for spreading false info.  My learning about BPF has been without mentorship, and due to my struggles I don't want others to have to fight through things I've already learned.  Unfortunately this makes me too eager to pipe up when I should let others like yourself speak first.  Again, thank you for sharing your knowledge, you have no idea how much it means to the community.

**Comment by anakryiko** [comment]
> @anakryiko Thank you for helping to educate us. BPF has been lacking some good documentation, and your recent posts have helped spread some good knowledge. EG I had no knowledge of bpf_core_field_exists() until yesterday.
> 
> A lot of my suggestion to use older kernels rather than newer kernels comes to missing bpf-helper functions and missing features such as bpf_ringbuffers. And of course, the dread glibc issues (though containerization takes this pain away). It's painful to develop a bpf application on Kernel 5.10 using features such as ringbuffers and bpf_d_path, linking against glibc 2.33, to then realize the code needs to run on 5.4 glibc 2.28.

As for the recommendation to go with newest vmlinux.h is at least partially because if you want to take advantage of some newer BPF feature in kernel (like newer map, newer BPF helpers, etc), having all the definitions in vmlinux.h is useful because you just can straight use those CO-RE feature-checking helper (field_exists, type_exists, enumval_exists) to decide what you do. If you use older vmlinux.h, then you'll have to define your own copies of those enums and types just to check if the running kernel supports it.

It's so convenient to do feature detection with BPF CO-RE, btw, that for complicated cases I chose to write and run once a special BPF object just to detect what's supported by the kernel (instead of doing detection from user-space). Check https://github.com/anakryiko/retsnoop/blob/master/src/calib_feat.bpf.c as an example. That calib_feat.bpf.c is separate from the main logic of retsnoop (https://github.com/anakryiko/retsnoop/blob/master/src/retsnoop.bpf.c statically linked together with https://github.com/anakryiko/retsnoop/blob/master/src/mass_attach.bpf.c), it's used purely for feature detection at the start up.

But it's up to everyone which approach they want to take. In practice you should always check which features map to which kernels before you decide to use some new BPF feature or map. So old or new vmlinux.h, you shouldn't be surprised.

glibc is completely unrelated to BPF side of things, so I can't comment on that.

> 
> IMHO it's easier to get the program going on 5.4 and then pull features into the app for targetting newer kernels/distros.
> 
> > It's just simply not true.
> 
> I apologize for spreading false info. My learning about BPF has been without mentorship, and due to my struggles I don't want others to have to fight through things I've already learned. Unfortunately this makes me too eager to pipe up when I should let others like yourself speak first. Again, thank you for sharing your knowledge, you have no idea how much it means to the community.

No worries and it's understandable. BPF and libbpf are still moving fast and best way to keep up is to be subscribed to bpf@vger.kernel.org mailing list. For documentation, I agree there can never be enough and good enough documentation, but in practice not many people are willing to dedicate their time and effort to improve the documentation. And for things that we do have documented, most people are too impatient to read it thoroughly. So in the end it all comes down to answering and explaining same things many times over, just with different people every time.

Anyways, as far as BPF CO-RE goes, it was designed and developed with real-world complex BPF application in mind. So far we (Facebook) haven't had a case where we couldn't accommodate kernel changes and features inside the single, compiled-once, BPF application. If you do run into such case, please send and email to bpf@vger.kernel.org and let's have a discussion.

**Comment by mimullin-bbry** [comment]
> but in practice not many people are willing to dedicate their time and effort to improve the documentation.

I do youtube videos where I attempt to document my learnings about Linuxy stuff.  A considerable amount of my content is around introductory BPF concepts.  I do this because I find audio and video are better for my brain to digest than written text, and I assume there are other people like me.

With your permission, I've learned a great deal from this discussion with you @anakryiko and @DevasiaThomas.  With your permission I'd like to use your blogpost "CORE reference guide" as a jumping off point for an upcoming video.

**Comment by DevasiaThomas** [comment]
@mimullin-bbry I'm in the same boat as you. I'm learning through examples and questions that everyone answers here and examples from regular libbpf in bcc and some tracee code.

**Comment by anakryiko** [comment]
> With your permission, I've learned a great deal from this discussion with you @anakryiko and @DevasiaThomas. With your permission I'd like to use your blogpost "CORE reference guide" as a jumping off point for an upcoming video.

Of course, I didn't write it for myself :) Please ping me with the URL when you have it ready, I'd be curious to check it out. Thanks.
